### ISSUE TITLE
exp: iterated SFT and rejection fine-tuning for multi-hypothesis contact synthesis

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#281]
## Kind

- Kind: models

## Question

Can prolonged, iterated supervised fine-tuning teach a contacts-v1 model to generate multiple useful structural hypotheses and synthesize a more accurate final contact prediction, and can subsequent rejection fine-tuning improve the hypothesis histories that support that synthesis?

## Hypothesis

1. An explicit `<final-prediction>` token separates hypothesis generation from synthesis, allowing the model to learn a distinct final-answer behavior.
2. After format warm-up, downweighting hypothesis-token loss (initial candidate: 0.1 relative to 1.0 for reference final contacts) will emphasize synthesis while maintaining hypothesis generation. This replaces the initially considered strict zero hypothesis loss.
3. Large offline datasets and repeated generation/SFT rounds permit a substantially longer optimization horizon than the previous online RL runs. Tens of thousands of optimizer steps are a horizon to investigate, not an established requirement; unique proteins and supervised target-token exposure must also be reported.
4. Training both complete and randomly interrupted histories will support natural emission of `<final-prediction>` and externally triggered finalization at different inference budgets.
5. Once synthesis works, selecting histories whose generated final answers score best, then applying nonzero SFT loss to those histories, can favor useful hypothesis-generation strategies. Complementary hypotheses may be selected when they improve synthesis, but increased diversity is not guaranteed: selection could also concentrate the sampling distribution.

## Background

- #163 established multi-hypothesis generation and investigated hypothesis/final loss weighting. Zero-hypothesis-loss training did not learn multiple sections; external-draft conditioning also degraded answer quality in that setup. Those results did not test the proposed dedicated final marker, two-stage warm-up/synthesis training, and long iterative schedule together.
- #230 transferred the multi format to a stronger base. Both #163 and #230 already ended training documents with ground-truth contacts. The changes here are explicit final-answer semantics, staged loss weighting, scale, repeated corpus refresh, and later trajectory selection. In #230, candidate voting beat oracle selection of an individual candidate, while independent plain sampling beat multi-section sampling at matched candidate count.
- #208 and #237 motivate revisiting SFT: their RL runs showed a limited useful optimization window and failures involving coverage, diversity, and trajectory length. Their results do not establish a general limit on RL or prove that longer SFT will succeed.
- #225 / #232 / #245 provide decontamination and evaluation conventions. Pin the actual starting checkpoint before generating data; do not silently inherit an older checkpoint just because it already supports the multi format.
- Related work: [STaR](https://arxiv.org/abs/2203.14465) and [ReST-EM](https://arxiv.org/abs/2312.06585) use iterative generation, selection, and fine-tuning. Our proposed main rejection variant retains reference contacts as the final training target rather than imitating the selected sampled answer.

### Document format and inference behavior

```text
<contacts-v1.multi>
<begin_sequence> ...sequence...
<begin_statements> ...hypothesis 1 contact triples...
<begin_statements> ...hypothesis 2 contact triples...
...
<final-prediction> ...final contact triples... <end>
```

During supervised training the final contact triples come from the reference structure. Hypotheses come from model samples generated without access to reference contacts. The final answer may combine or reject draft contacts and introduce contacts absent from all hypotheses; it is not restricted to selecting a candidate or taking their union.

Support both:

- Autonomous generation of hypotheses followed by natural emission of `<final-prediction>`.
- External insertion of `<final-prediction>` at a randomly sampled token budget, rounded to a complete contact-statement boundary. Interruption may occur partway through a hypothesis, but never within a `<contact> <pi> <pj>` triple. Mid-triple interruption is explicitly out of scope.

Reserve enough context for the final answer before allocating the hypothesis budget. Train across short and long prefixes, including no hypotheses and partial hypotheses. Do not silently truncate reference targets to make room for additional drafts.

### Phase A: format warm-up

Fine-tune the chosen current checkpoint to produce the multi format, the new final marker, and a properly terminated final prediction. Supervise hypotheses and structural transitions sufficiently to acquire the behavior, then verify free-running multi-section generation and natural finalization before beginning the synthesis-heavy phase. Include ordinary contacts-v1 rehearsal as needed to preserve plain-mode behavior; measure its effective supervised-weight share.

### Phase B: iterated synthesis SFT

1. Freeze a checkpoint for dataset construction.
2. Generate a large corpus of sequential multi-hypothesis histories from that checkpoint, matching the intended inference distribution. A concatenation of independent plain rollouts is a separate comparison, not an equivalent substitute.
3. Construct complete and randomly interrupted examples, followed by reference final contacts.
4. Continue training with reduced hypothesis loss and full final-answer loss.
5. Evaluate, refresh histories from the updated checkpoint, and repeat for several rounds. A modest replay fraction of older histories is a candidate stabilization measure, not a settled hyperparameter.

Proposed initial loss profile after warm-up:

| Target span | Complete example | Externally interrupted example |
|---|---:|---:|
| Hypothesis tokens and ordinary draft transitions | 0.1 | 0.1 |
| Predicting `<final-prediction>` | 1.0 | 0.0 |
| Reference final contact tokens and `<end>` | 1.0 | 1.0 |

The externally inserted marker is context, not a supervised decision to stop at an arbitrary budget. Its following answer is fully supervised. Complete examples maintain the autonomous stopping behavior. A 50:50 complete/interrupted mix and a small 0.03-vs-0.1 hypothesis-weight comparison are candidate initial settings, to finalize before launch.

Express masks in terms of the token being predicted and verify next-token alignment at boundaries. Normalize by supervised weight and report actual hypothesis/final/rehearsal weight totals: 0.1 per token is not a small total contribution if hypotheses are much longer than final answers. Zero direct token loss does not imply detaching context representations.

### Phase C: rejection fine-tuning after synthesis has begun to work

For each protein, sample several **whole trajectories**, each with its own hypothesis history and generated final answer. Proposed initial candidate count: four.

```text
Generate: hypotheses -> <final-prediction> -> sampled final answer
Select:   score sampled final answer against reference contacts
Train:    selected hypotheses -> <final-prediction> -> reference contacts
```

Select the trajectory with the best generated final answer, retain reduced but nonzero loss on its hypotheses, and retain full reference supervision after the final marker. Sampling several final answers from one shared history would not select among hypothesis-generation strategies and is not the main design. Training on the winning sampled final answer itself is a separate optional ablation, not the default.

For interrupted examples, sample one budget for the protein and apply it to all competing trajectories. Select using the answer generated at that interruption point, not a later completed answer. For autonomous examples, use a common maximum budget and report actual generation cost and natural stopping behavior.

Use final contact-set F1 as the initial candidate selection score, balancing precision and recall rather than rewarding precision alone. Lock the exact score, invalid/incomplete-output handling, and tie-breaking before launch. Do not score the ground-truth replacement itself or select by the best individual draft. Reference contacts are used for offline scoring and final targets only, never in generation prompts. Log winner and candidate-score distributions so selection on decoding luck or narrow strategies is visible.

### Controls and evaluation

Core comparisons, staged to avoid an unnecessarily large initial sweep:

- Direct reference-contact SFT without hypotheses, with the same protein pool and comparable final-target exposure.
- Synthesis SFT on a fixed corpus versus refreshed histories.
- Rejection best-of-four versus random-one-of-four selection from identically generated candidate pools, with matched training exposure.
- Voting over the supplied hypotheses and independent plain sampling at matched inference token budgets, also reporting wall time.

Diagnostics:

- Final-answer performance versus available hypothesis budget, including zero hypotheses, one hypothesis, multiple hypotheses, and repeated copies of one hypothesis.
- Old/new synthesizers evaluated on identical saved histories, alongside evaluation on their own fresh histories, to distinguish reading improvements from generator improvements.
- Final contact precision, recall, and F1; established R-precision/AUC using a preregistered final-output ranking/aggregation protocol rather than treating an unordered single contact set as a calibrated ranking.
- Hypothesis quality, true-contact union coverage, pairwise overlap, section counts, natural marker placement, format validity, completion rates, and plain-mode retention. Low overlap alone is not evidence of useful diversity.
- Paired per-protein comparisons and uncertainty, with candidate budget and target exposure explicit. Use eval-val for iteration, eval-denovo separately, and reserve eval-test for a settled configuration under #245's read-budget policy.

## Success criteria

- Format gate: at least 99% valid final-answer completion on the preregistered format-validation sample, measured separately for natural and forced finalization. Multi generation must retain multiple hypotheses in the intended multi-budget setting; immediate finalization alone does not pass the gate. Externally forced markers occur only between complete statements.
- Synthesis: final-answer accuracy improves over the no-hypothesis SFT control on eval-val, with a paired 95% confidence interval excluding zero on the preregistered primary metric. Report comparisons to voting and independent sampling at matched inference cost even if they remain stronger.
- Iteration: establish whether refreshed histories improve over fixed-corpus training at comparable target exposure; a negative result is informative and must be reported.
- Rejection: best-of-four selection improves over random-one-of-four training under the matched protocol. Report within-trajectory useful coverage and across-trajectory diversity separately; an accuracy gain alone does not establish diversity improvement.
- Freeze exact evaluation budgets, primary metric protocol, plain-mode retention tolerance, and checkpoint-selection rules before launch. Separate demonstrated effects from unconfirmed hypotheses.

## Notes

This issue records the design before implementation; no training or data-generation job has been launched for it.

Still to pin before launch: base checkpoint/tokenizer, eligible decontaminated protein pool and source mixture, corpus sizes, optimizer and total supervised-token budget, refresh schedule, context allocation, complete/interrupted mixture, and compute placement. Distinguish experimental structure references from predicted-structure labels in data provenance and reporting.

Implementation must follow repository conventions: experiment branch and PR, W&B routing to open-athena/MarinFold, run-history records, per-input prediction timing CSVs, region-co-located storage, and checkpoints saved with run name, step, and tokenizer. Publish reproducible corpus/checkpoint manifests and small comparison artifacts. Implementation begins as a subsequent task after this issue is filed.

**Comment by timodonnell** [comment]
🤖 The first full-model format trial has completed: 256 optimizer steps on 8×H100 in `cw-us-east-02a`, using 2,023 training and 25 held-out proteins. The current exp232 checkpoint and all large I/O stayed in the existing regional S3 bucket.

**The preregistered format gate failed.** At step 256, eight completions per held-out protein gave:

| Mode | Valid completions | Required |
| --- | ---: | ---: |
| Natural finalization | 0/200 (0%) | 99% |
| Forced finalization | 14/200 (7%) | 99% |

188/200 natural outputs contain multiple raw hypothesis-section markers, but only 8/200 emit the final marker. In forced mode, 185/200 restart a hypothesis after the final marker. The model is learning repeated sections but has not acquired reliable finalization. Invalid outputs remain in the denominator; none were filtered away to improve the result.

Validation loss fell from 2.7812 at step 32 to 2.2445 at 256. Peak allocation was 35.45 GB/GPU. A checkpoint-upload stall required resuming from complete step 64; that attempt saved through step 256. The README records the remaining upload-timeout risk and the correction for W&B suppressing replayed metrics.

[W&B run](https://wandb.ai/open-athena/MarinFold/runs/exp281-trial-s03) · [Results and reproduction](https://github.com/Open-Athena/MarinFold/blob/90fe7568/experiments/exp281_models_iterated_sft_and_rejection_fine_tuning/README.md) · [Draft implementation PR #283](https://github.com/Open-Athena/MarinFold/pull/283).

The public HF report, per-input timings, scores, diagnostics, and source manifests are preserved. All 16 pipeline tests pass. No synthesis/rejection training or FoldBench accuracy comparison has run. Next: extend format warm-up toward the proposed 2,000 steps with separate final-marker/termination diagnostics. This short pilot does not test the intended long SFT horizon, and #281 remains open.

**Comment by timodonnell** [comment]
🤖 Continuing the format warm-up as [exp281-format-s02](https://wandb.ai/open-athena/MarinFold/runs/exp281-format-s02), Iris `/bizon/exp281-format-s02`, source `4721c76c`.

The run preserves step-256 weights, Adam state, per-rank data position and RNG, and trains 1,744 additional steps to global step 2,000 on the same corpus and eight-H100 node profile. LR rewarms from 10% of peak over 100 steps, then decays; this is a schedule extension, not a from-scratch 2,000-step comparison. The final format gate is unchanged.

New teacher-forced diagnostics show why aggregate loss was misleading: at the starting checkpoint, natural-final-marker loss is 6.585 with 0/6 top-1 correct; multi-document end loss is 4.273 with 0/9 correct, while overall loss is 2.2445. These small-count diagnostics will be tracked alongside the final 400-sample evaluation.

Checkpoint publication now has a killable-process deadline of 300 seconds; completed checkpoints cannot be overwritten. W&B retains replayed measurements after rollback. All 19 tests and real CPU/GPU continuation-resume checks passed. Large I/O stays co-located on cw-us-east-02a; Iris showed 118 free H100s at placement time.

**Comment by timodonnell** [comment]
🤖 The extended format warm-up reached global step 1,750, then failed while publishing its checkpoint. The new 300-second deadline terminated the stalled publisher promptly; saves at 500, 750, 1,000, 1,250 and 1,500 had each completed in 97–101 seconds. Cleanup of the failed attempt's abandoned uploads succeeded.

The partial result shows pronounced overfitting of the frozen 2,023-document pilot corpus: training minibatch loss reaches 0.0024 while held-out loss rises from 2.2445 at step 256 to 5.9468. Natural final-marker and final-answer termination top-1 accuracies remain zero in the sparse teacher-forced checks (six natural markers and nine end tokens). This does not test larger refreshed SFT corpora or rejection training, and it does not replace the planned free-running format gate.

`/bizon/exp281-format-s02-r1` is queued in US-EAST-02A to resume the same run from complete step 1,500, with the exact training source/configuration (`4721c76c`). The latest capacity check showed no free H100s in US-EAST-02A and 416 in RNO2A. An RNO2A recovery is prepared but unsubmitted pending explicit approval for approximately 53 GB of cross-region checkpoint I/O.

Step 2,000 and its 400-completion natural/forced evaluation are still pending. No synthesis or rejection training has started. The README, committed partial history/CSV, transition/loss plots and run history record this state in PR #283.

[Continuation run on W&B](https://wandb.ai/open-athena/MarinFold/runs/exp281-format-s02).

**Comment by timodonnell** [comment]
🤖 The approved RNO2A recovery, step-2,000 checkpoint, and full format evaluation are complete.

The continuation still fails the preregistered format gate:

| Diagnostic | Step 256 | Step 2,000 |
| --- | ---: | ---: |
| Natural valid completions | 0/200 | 0/200 |
| Forced valid completions | 14/200 (7%) | 4/200 (2%) |
| Natural outputs containing `<final-prediction>` | 8/200 | 98/200 |
| Held-out weighted token loss | 2.2445 | 6.0468 |

Final training minibatch loss is 0.00117. The same small frozen corpus is strongly overfitting. Increased final-marker emission has not become reliable complete-document syntax: natural failures include incomplete/malformed contact triples, and 136/200 forced outputs fail with an unexpected hypothesis-start token. No valid natural trajectory has two nonempty hypotheses. This does not test the proposed larger refreshed SFT corpora or rejection training.

The east-region recovery had never begun training and was cancelled before the approved move. `/bizon/exp281-format-s02-rno-r1` resumed complete step 1,500 on eight H100s and successfully published steps 1,750 and 2,000 in 218.6 seconds each. Recovery worker duration: 1,349 seconds. Training source/configuration remained unchanged (`4721c76c`).

The final generation and scoring job used one RNO2A H100, shared one 5.886 GB model download across both modes, and finished in 472 seconds. The independent audit verifies all 400 candidates, 25 targets per mode, agreement with the scoring CSVs, and all 50 captured timing rows.

The 2.19 MB report is public at `hf://buckets/open-athena/MarinFold/data/exp281/format-s02/report.zip`, also attached as `open-athena/MarinFold/exp281-format-s02-report:v0`. Anonymous download checksum verification passed. PR #283 contains the updated README, raw history, per-input metrics/timings, plots, checkpoint provenance and recovery record.

No synthesis or rejection round was launched because the format gate failed. A larger and more varied format corpus, shorter initial histories, or stronger transition supervision are candidates for the next design; none is established as a remedy by this run.

[Completed continuation on W&B](https://wandb.ai/open-athena/MarinFold/runs/exp281-format-s02).
